dd-sdk-android-core/src/main/kotlin/com/datadog/android/core/internal/persistence/datastore/FileDataStoreHandler.kt

```diff
@@ -0,0 +1,289 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.android.core.internal.persistence.datastore
+
+import android.text.format.DateUtils
+import androidx.annotation.WorkerThread
+import com.datadog.android.api.InternalLogger
+import com.datadog.android.core.internal.persistence.Deserializer
+import com.datadog.android.core.internal.persistence.datastore.ext.toByteArray
+import com.datadog.android.core.internal.persistence.datastore.ext.toInt
+import com.datadog.android.core.internal.persistence.datastore.ext.toLong
+import com.datadog.android.core.internal.persistence.file.FileReaderWriter
+import com.datadog.android.core.internal.persistence.file.createNewFileSafe
+import com.datadog.android.core.internal.persistence.file.deleteSafe
+import com.datadog.android.core.internal.persistence.file.existsSafe
+import com.datadog.android.core.internal.persistence.file.mkdirsSafe
+import com.datadog.android.core.internal.persistence.tlvformat.FileTLVBlockReader
+import com.datadog.android.core.internal.persistence.tlvformat.TLVBlock
+import com.datadog.android.core.internal.persistence.tlvformat.TLVBlockType
+import com.datadog.android.core.persistence.Serializer
+import java.io.File
+import java.util.Locale
+
+@Suppress("TooManyFunctions")
+internal class FileDataStoreHandler(
+    private val sdkInstanceId: String,
+    private val storageDir: File,
+    private val internalLogger: InternalLogger,
+    private val fileReaderWriter: FileReaderWriter,
+    private val fileTLVBlockReader: FileTLVBlockReader,
+    private val dataStoreFileHelper: DataStoreFileHelper
+) : DataStoreHandler {
+    @WorkerThread
+    override fun <T : Any> write(
+        dataStoreFileName: String,
+        featureName: String,
+        serializer: Serializer<T>,
+        data: T
+    ) {
+        val dataStoreDirectory = createDataStoreDirectoryIfNecessary(featureName)
+        val dataStoreFile = createDataStoreFileIfNecessary(dataStoreDirectory, dataStoreFileName)
+
+        val lastUpdateBlock = getLastUpdateDateBlock()
+        val versionCodeBlock = getVersionCodeBlock()
+        val dataBlock = getDataBlock(data, serializer)
+
+        if (lastUpdateBlock == null || versionCodeBlock == null || dataBlock == null) return
+
+        writeToFile(
+            dataStoreFile,
+            lastUpdateBlock + versionCodeBlock + dataBlock
```

Comment:
performance nit: pretty sure that this operation will be like:

* create a byte array in the memory to host result of `lastUpdateBlock + versionCodeBlock`
* create a byte array in the memory to host result of above and ` + dataBlock`.

So to have a final result we will have 2 arrays/allocations (unless some very smart compiler optimization kicks in). Instead you can use this method https://github.com/DataDog/dd-sdk-android/blob/develop/dd-sdk-android-core/src/main/kotlin/com/datadog/android/core/internal/utils/ByteArrayExt.kt#L48 - it will make a single allocation.

```suggestion
            listOf(lastUpdateBlock, versionCodeBlock, dataBlock).join(separator = emptyByteArray())
```